src/Nerdbank.MessagePack/SecureHash/SecureEnumerableEqualityComparer`2.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Copyright (c) Andrew Arnott. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.InteropServices;
+
+namespace Nerdbank.MessagePack.SecureHash;
+
+/// <summary>
+/// A secure equality comparer that performs deep by-value comparisons, and uses collision-resistant hashing.
+/// </summary>
+/// <typeparam name="TEnumerable">The type of the enumerable.</typeparam>
+/// <typeparam name="TElement">The type of element in the enumerable.</typeparam>
+/// <param name="equalityComparer">The equality comparer for individual elements.</param>
+/// <param name="getEnumerable">The function that gets the enumerable of the collection.</param>
+internal class SecureEnumerableEqualityComparer<TEnumerable, TElement>(
+	SecureEqualityComparer<TElement> equalityComparer,
+	Func<TEnumerable, IEnumerable<TElement>> getEnumerable) : SecureEqualityComparer<TEnumerable>
+{
+	/// <inheritdoc/>
+	public override bool Equals(TEnumerable? x, TEnumerable? y)
+	{
+		if (x is null || y is null)
+		{
+			return ReferenceEquals(x, y);
+		}
+
+		IEnumerable<TElement> enumerableX = getEnumerable(x);
+		IEnumerable<TElement> enumerableY = getEnumerable(y);
+
+		if (Enumerable.TryGetNonEnumeratedCount(enumerableX, out int countX) &&
+			Enumerable.TryGetNonEnumeratedCount(enumerableY, out int countY) &&
+			countX != countY)
+		{
+			return false;
+		}
+
+		using IEnumerator<TElement> enumeratorX = enumerableX.GetEnumerator();
+		using IEnumerator<TElement> enumeratorY = enumerableY.GetEnumerator();
+		while (enumeratorX.MoveNext())
+		{
+			if (!enumeratorY.MoveNext() || !equalityComparer.Equals(enumeratorX.Current, enumeratorY.Current))
+			{
+				return false;
+			}
+		}
+
+		return !enumeratorY.MoveNext();
+	}
+
+	/// <inheritdoc/>
+	public override long GetSecureHashCode([DisallowNull] TEnumerable obj)
+	{
+		IEnumerable<TElement> enumerable = getEnumerable(obj);
+
+		// Ideally we could switch this to a SIP hash implementation that can process additional data in chunks with a constant amount of memory.
```

Comment:
This does look like a fairly expensive way for combining hashes.

Reply:
Are you talking about the fact that I have to allocate a list, or that I have to hash each individual member?

The allocation can theoretically be removed once we have a different SIP hash implementation.

We could also replace hashing of each element with hashing all the elements together if the following conditions hold:
1. The elements are all `unmanaged`.
2. The enumerable supports the `CollectionMarshal.GetSpan` method (or whatever it's called)
3. The element type does not implement `IDeepSecureEqualityComparer` or otherwise does not require normalization of its values.

Reply:
> Are you talking about the fact that I have to allocate a list, or that I have to hash each individual member?

The allocation, which I presumes stems from a requirement that all constituent hashes must be passed as inputs?

Reply:
Correct. Internally, the SIP hash implementation I have just iterates over the span in 64-bit blocks though, so it seems likely that I could alter the implementation myself to allow for simply iterating over the enumerable and incrementally hashing each element without ever allocating anything to store intermediate hashes if I can't find an implementation that already does this.

Reply:
Is there any way I can create an `unmanaged` optimized implementation and actually create it from your visitor class, without using reflection? C# won't let me because the T that's passed to the visitor doesn't have that constraint.